test/ContentValidationListenerTest.php

```diff
@@ -2095,6 +2096,104 @@ public function testSaveUnknownDataWhenEmptyInputFilter()
         $this->assertEquals($params, $dataParams->getBodyParams());
     }
 
+    public function testFilterEmptyEntriesFromDataByOptionWithNestedData()
```

Comment:
There are five different paths through the `removeEmptyData()` method, corresponding to five different conditions that can occur. You'll need five tests to fully exercise the method and verify it.

Reply:
Will have another look. Was thinking today it wasn't enough, but as tests passed and I'm pushed for time, I was thinking of coming back to it.